### ISSUE TITLE
Publish Docker image as a repository artifact

### ISSUE THREAD

**Issue by fkorotkov** [squidfunk/mkdocs-material#2031]
- [x] I've read the [contribution guidelines][1] and agree with them
- [x] ... to my best knowledge, my idea wouldn't break something for other users
- [x] ... the documentation does not mention anything about my idea
- [x] ... there are no open or closed issues that are related to my idea

## Description

It will be great to publish a Docker image as a package of this repository. 

### Use Cases

It will be easier for sponsors to consume the insiders version since GitHub will require authentication in order to pull the image so only sponsors will be able to pull it.

### Screenshots / Mockups

This action should work perfectly for both this repository and the insiders one thanks to `${{ github.event.repository.full_name }}`:

```yaml
name: Publish Docker image
on:
  release:
    types: [published]
jobs:
  push_to_registry:
    name: Push Docker image to GitHub Packages
    runs-on: ubuntu-latest
    steps:
      - name: Check out the repo
        uses: actions/checkout@v2
      - name: Push to GitHub Packages
        uses: docker/build-push-action@v1
        with:
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}
          registry: docker.pkg.github.com
          repository: ${{ github.event.repository.full_name }}/mkdocs
          tag_with_ref: true
```

  [1]: https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md

**Comment by squidfunk** [comment]
Great suggestion, and thanks for providing everything the implementation! We're currently pushing to Docker Hub, but I'm open to cutting down / refactoring the current build pipeline. It's a little messy after all. I'll come back to this asap.

**Comment by squidfunk** [comment]
So I've played around with it, but realized that there's a new version of the [`docker/build-push-action`](https://github.com/docker/build-push-action), namely v2. Could we update the recipe to the new version?

**Comment by fkorotkov** [comment]
It will look like this for the `v2` version:

```yaml
name: Publish Docker image
on:
  release:
    types: [published]
jobs:
  push_to_registry:
    name: Push Docker image to GitHub Packages
    runs-on: ubuntu-latest
    steps:
      - name: Check out the repo
        uses: actions/checkout@v2
      - name: Build
        uses: docker/build-push-action@v2
        with:
          tags: docker.pkg.github.com/${{ github.event.repository.full_name }}/mkdocs:${{ github.event.release.tag_name }}
      - name: Login to DockerHub
        uses: docker/login-action@v1 
        with:
          registry: docker.pkg.github.com
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}
      - name: Push to GitHub Packages
        run: docker push docker.pkg.github.com/${{ github.event.repository.full_name }}/mkdocs:${{ github.event.release.tag_name }}
```

I initially took the example from the official GH docs: https://docs.github.com/en/free-pro-team@latest/actions/guides/publishing-docker-images#publishing-images-to-github-packages

**Comment by squidfunk** [comment]
Awesome, thanks so looking into this! I'll try to address it this weekend.

**Comment by squidfunk** [comment]
I took the opportunity and refactored the whole build pipeline into multiple workflows. Futhermore, I decided to use GitHub Container Registry, as Docker support is going to be deprecated next year for GitHub Package Registry. Thus, to use it, you must opt-in to the public beta of GHCR. The official image is now dual-released on DockerHub and GHCR:

```
docker pull ghcr.io/squidfunk/mkdocs-material 
```

Insiders should also be ready in a few hours.
